frontends/PyCDE/src/pycde/esi.py

```diff
@@ -169,6 +177,49 @@ def Cosim(decl: ServiceDecl, clk, rst):
   decl.instantiate_builtin("cosim", [], [clk, rst])
 
 
+def CosimBSP(user_module):
+  """Wrap and return a cosimulation 'board support package' containing
+  'user_module'"""
+  from .module import module, generator
+  from .common import Clock, Input
+
+  @module
+  class top:
+    clk = Clock()
+    rst = Input(types.int(1))
+
+    @generator
+    def build(ports):
+      user_module(clk=ports.clk, rst=ports.rst)
+      raw_esi.ServiceInstanceOp(result=[],
+                                service_symbol=None,
+                                impl_type=ir.StringAttr.get("cosim"),
+                                inputs=[ports.clk.value, ports.rst.value])
+
+      System.current().add_packaging_step(top.package)
+
+    @staticmethod
+    def package(sys: System):
+      """Run the packaging to create a cosim package."""
+      # TODO: this only works in-tree. Make it work in packages as well.
+      build_dir = __dir__.parents[4]
```

Comment:
By "in-tree", I meant the "in-tree" build directory rather than the installed package. I'm not even sure that these files are present in the python package. We don't have a solid "install" target for pycde, so I'm not sure how to test for this other than asserting that the name matches, which I'll do. As of today, the wheel just copies the python dir: https://github.com/llvm/circt/blob/062eab31a67c8a8418b2ba9fb4f1814cdca08783/frontends/PyCDE/setup.py#L89 which ain't great.

Reply:
#4344